providers/openstack/zun.go

```diff
@@ -0,0 +1,509 @@
+			ClusterName:       capsule.MetaLabels["ClusterName"],
+			UID:               types.UID(capsule.UUID),
+			CreationTimestamp: podCreationTimestamp,
+		},
+		Spec: v1.PodSpec{
+			NodeName:   capsule.MetaLabels["NodeName"],
+			Volumes:    []v1.Volume{},
+			Containers: containers,
+		},
+
+		Status: v1.PodStatus{
+			Phase:             zunCapStatusToPodPhase(capsule.Status),
+			Conditions:        []v1.PodCondition{},
+			Message:           "",
+			Reason:            "",
+			HostIP:            "",
+			PodIP:             ip,
+			StartTime:         &containerStartTime,
+			ContainerStatuses: containerStatuses,
+		},
+	}
+
+	return &p, nil
+}
+
+// UpdatePod is a noop, Zun currently does not support live updates of a pod.
+func (p *ZunProvider) UpdatePod(ctx context.Context, pod *v1.Pod) error {
+	return nil
+}
+
+// DeletePod deletes the specified pod out of Zun.
+func (p *ZunProvider) DeletePod(ctx context.Context, pod *v1.Pod) error {
+	return capsules.Delete(p.ZunClient, fmt.Sprintf("%s-%s", pod.Namespace, pod.Name)).ExtractErr()
+}
+
+func zunContainerStausToContainerStatus(cs *capsules.Container) v1.ContainerState {
+	// Zun already container start time but not add support at gophercloud
+	//startTime := metav1.NewTime(time.Time(cs.StartTime))
+
+	// Zun container status:
+	//'Error', 'Running', 'Stopped', 'Paused', 'Unknown', 'Creating', 'Created',
+	//'Deleted', 'Deleting', 'Rebuilding', 'Dead', 'Restarting'
+
+	// Handle the case where the container is running.
+	if cs.Status == "Running" || cs.Status == "Stopped" {
+		return v1.ContainerState{
+			Running: &v1.ContainerStateRunning{
+				StartedAt: metav1.NewTime(time.Time(cs.StartedAt)),
+			},
+		}
+	}
+
+	// Handle the case where the container failed.
+	if cs.Status == "Error" || cs.Status == "Dead" {
+		return v1.ContainerState{
+			Terminated: &v1.ContainerStateTerminated{
+				ExitCode:   int32(0),
+				Reason:     cs.Status,
+				Message:    cs.StatusDetail,
+				StartedAt:  metav1.NewTime(time.Time(cs.StartedAt)),
+				FinishedAt: metav1.NewTime(time.Time(cs.UpdatedAt)),
+			},
+		}
+	}
+
+	// Handle the case where the container is pending.
+	// Which should be all other Zun states.
+	return v1.ContainerState{
+		Waiting: &v1.ContainerStateWaiting{
+			Reason:  cs.Status,
+			Message: cs.StatusDetail,
+		},
+	}
+}
+
+func zunStatusToPodPhase(status string) v1.PodPhase {
+	switch status {
+	case "Running":
+		return v1.PodRunning
+	case "Stopped":
+		return v1.PodSucceeded
+	case "Error":
+		return v1.PodFailed
+	case "Dead":
+		return v1.PodFailed
+	case "Creating":
+		return v1.PodPending
+	case "Created":
+		return v1.PodPending
+	case "Restarting":
+		return v1.PodPending
+	case "Rebuilding":
+		return v1.PodPending
+	case "Paused":
+		return v1.PodPending
+	case "Deleting":
+		return v1.PodPending
+	case "Deleted":
+		return v1.PodPending
+	}
+
+	return v1.PodUnknown
+}
+
+func zunCapStatusToPodPhase(status string) v1.PodPhase {
```

Comment:
In overall, with the latest API version, container and capsule status would be the same, so I guess we can remove this function and re-use zunStatusToPodPhase for converting the capsule's status to pod.